processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -114,83 +134,82 @@ public boolean isInitialized()
-          fromKey.remaining(),
-          keySerde.keySize()
-      );
-    }
+    Preconditions.checkNotNull(key);
 
     final int recordOffset = dimIndex * recordSize;
 
-    if (recordOffset + recordSize > buffer.capacity()) {
+    if (recordOffset + recordSize > valBuffer.capacity()) {
       // This error cannot be recoverd, and the query must fail
       throw new ISE(
           "A record of size [%d] cannot be written to the array buffer at offset[%d] "
           + "because it exceeds the buffer capacity[%d]. Try increasing druid.processing.buffer.sizeBytes",
           recordSize,
           recordOffset,
-          buffer.capacity()
+          valBuffer.capacity()
       );
     }
 
-    if (!isUsedKey(buffer, recordOffset)) {
-      this.buffer.position(recordOffset);
-      this.buffer.putInt(Groupers.getUsedFlag(dimIndex));
-      this.buffer.put(fromKey);
+    if (!isUsedSlot(dimIndex)) {
+      initializeSlot(dimIndex);
     }
 
-    final int baseOffset = recordOffset + keySize;
     for (int i = 0; i < aggregators.length; i++) {
-      aggregators[i].aggregate(buffer, baseOffset + aggregatorOffsets[i]);
+      aggregators[i].aggregate(valBuffer, recordOffset + aggregatorOffsets[i]);
     }
 
     return AggregateResult.ok();
   }
 
-  @Override
-  public AggregateResult aggregate(KeyType key)
+  private void initializeSlot(int dimIndex)
   {
-    // BufferArrayGrouper is used only for dictionary-indexed single-value string dimensions.
-    // Here, the key contains the dictionary-encoded value of the grouping key
-    // and we use it as the index for the aggregation array.
-    final ByteBuffer fromKey = keySerde.toByteBuffer(key);
-    final int dimIndex = fromKey.getInt() + 1; // the first index is for missing value
-    fromKey.rewind();
-    return aggregate(key, dimIndex);
+    final int index = dimIndex / 8;
+    final int extraIndex = dimIndex % 8;
+    usedFlagBuffer.put(index, (byte) (usedFlagBuffer.get(index) | 1 << extraIndex));
+
+    final int recordOffset = dimIndex * recordSize;
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(valBuffer, recordOffset + aggregatorOffsets[i]);
+    }
   }
 
-  private static boolean isUsedKey(ByteBuffer buffer, int pos)
+  private boolean isUsedSlot(int dimIndex)
   {
-    return (buffer.get(pos) & 0x80) == 0x80;
+    final int index = dimIndex / 8;
+    final int extraIndex = dimIndex % 8;
+    final int usedByte = 1 << extraIndex;
+    return (usedFlagBuffer.get(index) & usedByte) == usedByte;
   }
 
   @Override
   public void reset()
   {
-    buffer.putInt(0, 0); // for missing value
-    for (int i = 1; i < cardinality + 1; i++) {
-      buffer.putInt(i * recordSize, 0);
-      final int baseOffset = i * recordSize + keySize;
-      for (int j = 0; j < aggregators.length; ++j) {
-        aggregators[j].init(buffer, baseOffset + aggregatorOffsets[j]);
-      }
+    final int usedBufferCapacity = usedFlagBuffer.capacity();
```

Comment:
I'd like to stick to as-is because ```reset()``` does only clearing the buffer. I added a comment what is going on here.